test/parallel_api/algorithm/alg.nonmodifying/transform_if.pass.cpp

```diff
@@ -0,0 +1,165 @@
+// -*- C++ -*-
+//===-- transform_if.pass.cpp ----------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "support/test_config.h"
+
+#include _PSTL_TEST_HEADER(execution)
+#include _PSTL_TEST_HEADER(algorithm)
+
+#include "support/utils.h"
+
+using namespace TestUtils;
+
+template <typename T1, typename T2>
+struct mutable_negate_first
+{
+    T1
+    operator()(const T1& a, const T2&) //explicitly not const
+    {
+        return -a;
+    }
+};
+
+template <typename T>
+struct mutable_negate
+{
+    T
+    operator()(const T& a) //explicitly not const
+    {
+        return -a;
+    }
+};
+
+template <typename T1, typename T2>
+struct mutable_check_mask_second
+{
+    bool
+    operator()(const T1&, const T2& b) //explicitly not const
+    {
+        return b == 1;
+    }
+};
+
+template <typename T>
+struct mutable_check_mod3_is_0
+{
+    bool
+    operator()(const T& a) //explicitly not const
+    {
+        return (a % 3) == 0;
+    }
+};
+
+struct test_transform_if_binary
+{
+    template <typename Policy, typename InputIterator1, typename InputIterator2, typename OutputIterator, typename Size>
+    void
+    operator()(Policy&& exec, InputIterator1 first, InputIterator1 last, InputIterator2 mask,
+               InputIterator2 /*mask_end*/, OutputIterator result_begin, OutputIterator result_end, Size n)
+    {
+        using in_value_type1 = typename ::std::iterator_traits<InputIterator1>::value_type;
+        using in_value_type2 = typename ::std::iterator_traits<InputIterator2>::value_type;
+        using out_value_type = typename ::std::iterator_traits<OutputIterator>::value_type;
+        // call transform_if
+        oneapi::dpl::transform_if(exec, first, last, mask, result_begin,
+                                  mutable_negate_first<in_value_type1, in_value_type2>{},
+                                  mutable_check_mask_second<in_value_type1, in_value_type2>{});
+
+        //calculate expected
+        std::vector<out_value_type> expected(n);
+        auto in_iter = first;
+        auto mask_iter = mask;
+        auto expected_iter = expected.begin();
+        for (; in_iter != last; in_iter++, mask_iter++, expected_iter++)
+        {
+            *expected_iter = *mask_iter == 1 ? -(*in_iter) : 0;
+        }
+
+        EXPECT_EQ_N(expected.begin(), result_begin, n, "wrong effect from transform_if binary");
+
+        // reset output elements to 0
+        ::std::fill(result_begin, result_end, 0);
```

Comment:
I believe this call is not required. Please check.

Reply:
This resets the output to its original state in between invocations of the test.  This seems correct given that `transform_if` preserves the initial value of the output range for elements where the predicate is `false`. 

Technically, the test would pass without it because the same locations in consecutive invocations of the test will have their predicate be `false`, but it seems dependent on the predicate and structure of testing.  

I've added comments and a non-zero `init_val` to make it more clear why we are doing this.

Reply:
ok